rfd/0180-sso-as-mfa-method.md

```diff
@@ -0,0 +1,765 @@
+  string request_id = 1;
+  // RedirectUrl is an IdP redirect URL to initate the SSO MFA flow.
+  string redirect_url = 2;
+}
+```
+
+```diff
+message MFAAuthenticateResponse {
+  oneof Response {
+    TOTPResponse TOTP = 2;
+    webauthn.CredentialAssertionResponse Webauthn = 3;
++   SSOResponse SSO = 4;
+  }
+}
+```
+
+**SSOResponse**
+
+```proto
+// SSOResponse is a response to SSOChallenge.
+message SSOResponse {
+  // RequestId is the ID of an SSO auth request.
+  string request_id = 1;
+  // MFAToken is a secret token used to verify the user's SSO MFA session.
+  string mfa_token = 2;
+}
+```
+
+```diff
+message MFAAuthenticateChallenge {
+  ...
++ // SSO Challenge is an sso MFA challenge. If set, the client can go to the
++ // IdP redirect URL to perform an MFA check in the IdP and obtain an MFA
++ // token. This token and sso request pair can then be used as MFA verification.
++ SSOChallenge SSOChallenge = 5;
+}
+```
+
+**SSOMFADevice**
+
+```proto
+// SSOMFADevice contains details of an SSO MFA method.
+message SSOMFADevice {
+  // connector_id is the ID of the SSO connector.
+  string connector_id = 1;
+  // connector_type is the type of the SSO connector.
+  string connector_type = 2;
+}
+```
+
+```diff
+message MFADevice {
+  ...
+  oneof device {
+    TOTPDevice totp = 8;
+    U2FDevice u2f = 9;
+    WebauthnDevice webauthn = 10;
++   SSOMFADevice sso = 11;
+  }
+}
+```
+
+**AuthConnectorMFAMode**
+
+```proto
+// AuthConnectorMFAMode specified the MFA mode of an Auth Connector.
+enum AuthConnectorMFAMode {
+  // UNSPECIFIED is treated as OPTIONAL.
+  AUTH_CONNECTOR_MFA_MODE_UNSPECIFIED = 0;
+  // OPTIONAL mfa through this auth connector is optional and other MFA methods are preferred.
+  AUTH_CONNECTOR_MFA_MODE_OPTIONAL = 1;
+  // PREFERRED mfa through this auth connector is preferred and other MFA methods are optional.
+  AUTH_CONNECTOR_MFA_MODE_PREFERRED = 2;
+  // REQUIRED mfa through this auth connector is required and other MFA methods are disabled.
+  AUTH_CONNECTOR_MFA_MODE_REQUIRED = 3;
+}
+```
+
+**SAMLConnectorMFASettings**
+
+```proto
+message SAMLConnectorMFASettings {
+  // Enabled specified whether this SAML connector supports MFA checks. Defaults to false.
+  bool enabled = 1;
+  // MFAMode specifies what MFA mode this auth connector will operate with. Defaults to "optional".
+  AuthConnectorMFAMode mfa_mode = 2;
+  // EntityDescriptor is XML with descriptor. It can be used to supply configuration
+  // parameters in one XML file rather than supplying them in the individual elements.
+  // If unset, the parent SAML connector's EntityDescriptor will be used.
+  string entity_descriptor = 3;
+  // EntityDescriptorUrl is a URL that supplies a configuration XML. If unset, 
+  // the parent SAML connector's EntityDescriptor will be used.
+  string entity_descriptor_url = 4;
+  // ForceReauth specified whether re-authentication should be forced for MFA checks. UNSPECIFIED is 
```

Comment:
`ForceReauth` will always force the user to fully re-login. This would prevent admins from creating an MFA flow like I described before:

> Okta SAML App that skips the username/password prompt for existing sessions and just prompts for MFA.

We default to `ForceReauth=true` and `max_age=0` to protect against the least configured scenario (admin uses same SSO client for Login and MFA, MFA checks would be non-interactively checking for an active session, so we should force re-auth), while still allowing admins to change the value for fully configured use cases (MFA checks or at least device trust).

My apologies for leaving this nuance unclear, I'll try to cover all the scenarios I was looking to cover more clearly in a new section.

Reply:
I think it would be better if you add a table in RFD that maps required config, both on Teleport and in the IdP, that affects authentication strategy in the IdP with a known authentication outcome. 

For example: 
#### Configurations required for each IdP to always enforce re-authentication

| IdP Type | Teleport config | IdP Config |
| --- | --- | --- |
| Okta SAML | ForceReauth=true | any config needed on IdP side? session ttl? auth policy? |
|  | ForceReauth=false | any config needed on IdP side? session ttl? auth policy? |
| Okta OIDC | max_age=0 | any config needed on IdP side? session ttl? auth policy? |

You may not have a bandwidth to test this with all currently supported IdP but at least for Okta which is supposedly the primary IdP we want to support immediately?

That will also come in handy for users during implementation.

Reply:
I discussed this with @rosstimothy earlier. The issue is that we don't have access to Okta/Auth0/PingIdentity enterprise dev accounts, and MFA is an enterprise feature for each of these. I could follow the docs to layout a configuration that I believe would be good, but I wouldn't be able to test it. Is there another mfa-enabled IdP I should test with?

Reply:
The intention with the current design is to leave the specifics of the IdP MFA implementation to the IdP admins. We force re-auth by default because we have no way to determine from Teleport whether the IdP is properly set up with MFA checks or not. An admin should determine whether or not to disable forced re-authentication based on whether their IdP implementation has strict MFA/device checks or not.

Reply:
Got it. There isn't any specific mfa-enabled IdP I can suggest here. My reason for asking that is because each IdP may treat force-auth differently, depending on the application (service provider) specific session and authentication policy. So as a feature provider, it would be good to have us figure out proper config combination and recommend that to the admins. And initially I myself wasn't clear but your comment https://github.com/gravitational/teleport/pull/44699#discussion_r1752362571 helped me understand it. 

I guess there isn't alternative to what you suggested given lack of access to enterprise dev idP instances.